rpcs3/Emu/Cell/Modules/cellGem.cpp

```diff
@@ -304,6 +290,119 @@ static bool map_ext_to_ds3_input(const u32 port_no, CellGemExtPortData& ext)
 	return true;
 }
 
+/**
+ * \brief Maps Move controller data (digital buttons, and analog Trigger data) to mouse input.
+ *        Move Button: Mouse1
+ *        Trigger:     Mouse2
+ * \param mouse_no Mouse index number to use
+ * \param digital_buttons Bitmask filled with CELL_GEM_CTRL_* values
+ * \param analog_t Analog value of Move's Trigger.
+ * \return true on success, false if mouse mouse_no is invalid
+ */
+static bool mouse_input_to_pad(const u32 mouse_no, be_t<u16>& digital_buttons, be_t<u16>& analog_t)
+{
+	const auto handler = g_fxo->get<MouseHandlerBase>();
+
+	std::scoped_lock lock(handler->mutex);
+
+	if (!handler || mouse_no >= handler->GetMice().size())
+	{
+		return false;
+	}
+
+	memset(&digital_buttons, 0, sizeof(digital_buttons));
+
+	const auto& mouse_data = handler->GetMice().at(0);
+
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1 && mouse_data.buttons & CELL_MOUSE_BUTTON_2)
+		digital_buttons |= CELL_GEM_CTRL_CIRCLE;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_3)
+		digital_buttons |= CELL_GEM_CTRL_CROSS;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_2)
+		digital_buttons |= CELL_GEM_CTRL_MOVE;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1 && mouse_data.buttons & CELL_MOUSE_BUTTON_3)
+		digital_buttons |= CELL_GEM_CTRL_START;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_1)
+		digital_buttons |= CELL_GEM_CTRL_T;
+	if (mouse_data.buttons & CELL_MOUSE_BUTTON_2 && mouse_data.buttons & CELL_MOUSE_BUTTON_3)
+		digital_buttons |= CELL_GEM_CTRL_TRIANGLE;
+
+	analog_t = mouse_data.buttons & CELL_MOUSE_BUTTON_1 ? 0xFFFF : 0;
+
+	return true;
+}
+
+static bool mouse_pos_to_gem_image_state(const u32 mouse_no, vm::ptr<CellGemImageState>& gem_image_state)
+{
+		// Only game this seems to work on is PAIN, others use different functions
+
+		const auto handler = g_fxo->get<MouseHandlerBase>();
+
+		std::scoped_lock lock(handler->mutex);
+
+		if (!gem_image_state || !handler || mouse_no >= handler->GetMice().size())
+		{
+			return false;
+		}
+
+		const auto& mouse = handler->GetMice().at(0);
+
+		const auto renderer = static_cast<GSRender*>(rsx::get_current_renderer());
+		const auto width = renderer->get_frame()->client_width();
+		const auto hight = renderer->get_frame()->client_height();
+		const f32 scaling_width = width / 640.f;
+		const f32 scaling_hight = hight / 480.f;
```

Comment:
What are all these magic numbers in all the helper functions?

Reply:
is obligatory for scaling